### ISSUE TITLE
[feature] Add LLM dataset preparation workflows

### ISSUE THREAD

**Issue by liding-nv** [NVIDIA-NeMo/Megatron-Bridge#3918]
## Problem

Megatron-Bridge provides model recipes and data loaders, but it largely assumes users already have training-ready dataset artifacts. For LLM workflows, this leaves a gap between curated text/chat data and Bridge-consumable inputs for pretraining or SFT/LoRA.

Nemotron fills this gap with a last-mile data-preparation layer under `src/nemotron/data_prep`. This issue proposes documenting and porting the relevant LLM pieces into Megatron-Bridge.

## Scope

In scope:

- LLM text pretraining, sft, lora data preparation.
- Input discovery from local files, object storage, and Hugging Face dataset repos.
- Conversion from raw records into Bridge-consumable artifacts.
- Documentation of input schemas, output contracts, and recipe integration points.

Out of scope:

- RL/Omni data preparation.
- Raw-data curation pipelines.
- NeMo Curator integration.

Nemotron contains separate NeMo Curator-based curation recipes, but the LLM training data-prep path discussed here does not invoke NeMo Curator. Treat NeMo Curator as upstream curation for this issue.

## Current Bridge State

Bridge can already consume prepared data:

- Pretraining consumes Megatron `.bin/.idx` datasets through `data_path`, `blend`, `blend_per_split`, or `per_split_data_args_path`.
- SFT/LoRA can consume packed Parquet through `PackedSequenceSpecs`.
- Bridge has some local/HF SFT conveniences.

The gap is not runtime data loading. The gap is producing the expected artifacts from raw or curated datasets in a documented, reproducible, scalable way.

## Shared Input Contract From Nemotron

Nemotron's relevant LLM data-prep flows are file-oriented.

Supported sources: (1) local filesystem, (2) object storage through `fsspec`, such as S3/GCS, and (3) HF dataset repos via `hf://...`.

Supported source files:

- `.parquet`
- `.jsonl`
- line-delimited `.json`

This issue assumes those source files already exist. They may come from upstream curation, synthetic-data generation, manual dataset construction, or a public/private Hugging Face dataset export. Producing the raw text documents or OpenAI-style chat records is outside this issue; the Bridge gap starts at taking those records and materializing training-ready artifacts.

For Hugging Face datasets, Nemotron discovers matching data files in the dataset repo and downloads/resolves those files. This path is not primarily based on `datasets.load_dataset()`.

## Pretraining Data Prep

Nemotron pretraining prep converts raw text records into Megatron `.bin/.idx` shards.

### Input Schema

Each record needs a text field, default `text`.

```json
{"text": "Dataset preparation converts raw text into tokenized Megatron bin/idx files."}
```

This record is expected to have been produced upstream, for example by a data curation/export pipeline. The pretraining prep flow does not create the document text itself.

Extra columns may exist; the pretraining tokenizer stage reads the configured `text_field`.

### Step-by-Step Flow

1. User provides a `DataBlend` manifest with dataset name, path, weight, split/subset, and `text_field`.
2. Data prep resolves the tokenizer and records an immutable tokenizer revision/hash.
3. It builds a deterministic run config and computes a `run_hash`.
4. It discovers source files from local storage, object storage, or Hugging Face Hub.
5. It fingerprints source files using path, size, etag/revision/mtime where available.
6. It creates a deterministic shard plan, with row-level splitting when there are fewer files than requested shards.
7. Workers download or resolve assigned source files.
8. The tokenizer stage reads text records from Parquet/JSONL.
9. It extracts the configured text field and applies optional `min_doc_chars`, `max_doc_tokens`, and `max_rows`.
10. It tokenizes documents.
11. It writes Megatron `.bin/.idx` shard files and per-shard receipts.
12. Finalization scans receipts, aggregates stats, and emits weighted data prefixes or split-aware blend data for Bridge.

### Output Contract

Expected artifact shape:

```text
runs/<run_hash>/datasets/<dataset>/<plan_hash>/shard_000000.bin
runs/<run_hash>/datasets/<dataset>/<plan_hash>/shard_000000.idx
runs/<run_hash>/datasets/<dataset>/<plan_hash>/shard_000001.bin
runs/<run_hash>/datasets/<dataset>/<plan_hash>/shard_000001.idx
```

Bridge-facing split output can be represented as weighted prefixes:

```json
{
  "train": ["1.0", "/path/to/train_shard_prefix"],
  "valid": ["1.0", "/path/to/valid_shard_prefix"],
  "test": []
}
```

## SFT/LoRA Data Prep

Nemotron SFT prep converts OpenAI-style chat records into packed Parquet shards. LoRA uses the same prepared data as full SFT; PEFT changes training behavior, not the dataset format.

### Input Schema

Each record needs a chat messages field, default `messages`. The field is a list of OpenAI-style messages.

```json
{
  "messages": [
    {"role": "system", "content": "You are a concise assistant."},
    {"role": "user", "content": "What is 2 + 2?"},
    {"role": "assistant", "content": "2 + 2 is 4."}
  ]
}
```

Optional fields can include `tools`, `used_in`, or other metadata used for filtering.

These chat records are expected to have been produced upstream, for example by human annotation, synthetic-data generation, an instruction-data pipeline, or a dataset export. The SFT/LoRA prep flow does not create the conversation content itself; it formats, validates, tokenizes, masks, and packs existing conversations.

### Step-by-Step Flow

1. User provides a `DataBlend` manifest with dataset name, path, weight, split/subset, and chat field names.
2. Data prep resolves tokenizer and chat-template configuration.
3. It builds a deterministic run config and computes a `run_hash`.
4. It discovers, fingerprints, shards, and downloads/resolves source files.
5. The SFT stage reads chat records from Parquet/JSONL.
6. It optionally filters records using metadata such as `used_in`.
7. It validates and normalizes OpenAI-style conversations, including optional `tools`.
8. It applies the tokenizer chat template.
9. It builds role-aware loss masks: system/user tokens are masked out, assistant tokens contribute to loss.
10. It tokenizes examples into an intermediate sequence spool.
11. It packs tokenized examples up to `pack_size`.
12. Supported packing algorithms are `first_fit_shuffle`, `first_fit_decreasing`, and `concatenative`.
13. It records `seq_start_id` boundaries for original sequences inside each packed sample.
14. It shifts loss masks within each subsequence to match Bridge label semantics, with the last token of each subsequence masked out.
15. It writes packed Parquet shards, receipts, stats, and train/valid split directories or paths for `PackedSequenceSpecs`.

### Output Contract

Each packed Parquet shard contains:

```text
input_ids: list<int32>
loss_mask: list<uint8>
seq_start_id: list<int32>
```

Expected artifact shape:

```text
runs/<run_hash>/datasets/<dataset>/<plan_hash>/shard_000000.parquet
runs/<run_hash>/datasets/<dataset>/<plan_hash>/shard_000001.parquet
splits/train/*.parquet
splits/valid/*.parquet
```

Bridge should consume these through `PackedSequenceSpecs`:

```python
PackedSequenceSpecs(
    packed_sequence_size=<pack_size>,
    packed_train_data_path="/path/to/splits/train",
    packed_val_data_path="/path/to/splits/valid",
)
```

## Concrete Toy Examples

These examples are supporting material only. They are not intended to match an actual Nemotron dataset.

### Pretraining JSONL

```jsonl
{"text": "Megatron Bridge provides training recipes for large language models."}
{"text": "Dataset preparation converts raw text into tokenized Megatron bin/idx files."}
{"text": "This is another document that will become one training sequence before optional truncation."}
```

### Pretraining Manifest

```json
{
  "datasets": [
    {
      "name": "toy_pretrain",
      "path": "./data/pretrain/toy_pretrain.jsonl",
      "weight": 1.0,
      "text_field": "text"
    }
  ]
}
```

### SFT/LoRA JSONL

```jsonl
{"messages":[{"role":"system","content":"You are a concise assistant."},{"role":"user","content":"What is 2 + 2?"},{"role":"assistant","content":"2 + 2 is 4."}]}
{"messages":[{"role":"user","content":"Define sequence packing."},{"role":"assistant","content":"Sequence packing combines shorter examples into one fixed-length training sequence."}]}
```

### SFT/LoRA Manifest

```json
{
  "datasets": [
    {
      "name": "toy_sft",
      "path": "./data/sft/toy_sft.jsonl",
      "weight": 1.0
    }
  ]
}
```

## Gaps To Address

Pretraining gaps:

- No Bridge-owned `DataBlend`-style preparation contract.
- No Bridge-native source discovery, sharding, download, tokenization, `.bin/.idx` production, and finalization workflow.
- No deterministic run hash/source fingerprinting or receipt-based resume flow.
- No documented path from raw text records to Bridge pretraining config fields.

SFT/LoRA gaps:

- Bridge can read packed Parquet, but does not produce it.
- No Bridge-native chat-template, tokenize, mask, pack, and write workflow.
- Packing behavior and loss-mask semantics are not documented as a preparation contract.
- Users must externally produce files matching `PackedSequenceSpecs`.

Shared gaps:

- No standard raw input schema documentation.
- No standard output artifact layout.
- No small end-to-end examples showing raw records, preparation, and recipe consumption.
- No clear guidance on optional distributed execution dependencies.

## Recommended Porting Plan

Port the Nemotron LLM data-prep flow into Bridge in layers.

1. Define a Bridge-owned manifest and artifact contract.
   - Support dataset `name`, `path`, `weight`, `split`, `subset`, `text_field`, `messages_field`, `tools_field`, and optional filtering metadata.
   - Document `.parquet`, `.jsonl`, and line-delimited `.json` inputs.
   - Document `.bin/.idx` and packed Parquet outputs.

2. Port shared discovery and planning.
   - Local/object-store/HF dataset file discovery.
   - Deterministic source fingerprinting.
   - Deterministic shard planning.
   - Row-level splitting when file count is smaller than shard count.

3. Port the pretraining producer.
   - Read text records, extract `text_field`, filter/truncate, tokenize, write `.bin/.idx`.
   - Finalize weighted prefixes or per-split blend output for Bridge recipes.

4. Port the SFT/LoRA producer.
   - Read OpenAI-style `messages`, support optional `tools` and filters, validate conversations, apply chat template, create assistant-only loss masks, and tokenize examples.
   - Keep LoRA on the same prepared-data path as full SFT.

5. Port the packed-sequence writer.
   - Support `first_fit_shuffle`, `first_fit_decreasing`, and `concatenative`.
   - Preserve `input_ids`, `loss_mask`, and `seq_start_id`.
   - Preserve Bridge-compatible boundary and loss-mask semantics.

6. Port resumability and finalization.
   - Per-shard receipts, output verification, skip/resume behavior, final stats, stable run layout, and deterministic `run_hash`.

7. Add user-facing entry points.
   - Provide CLI/script/API surfaces such as `prepare-pretrain-data` and `prepare-sft-data`.
   - Show how outputs connect to Bridge fields: `data_path`, `blend`, `blend_per_split`, `per_split_data_args_path`, and `PackedSequenceSpecs`.

8. Keep distributed execution optional.
   - Core preparation logic should be backend-neutral.
   - Ray/Cosmos-Xenna or similar distributed backends should be optional extras, not required Bridge dependencies.

9. Add focused tests and examples.
   - Cover toy JSONL inputs, source discovery, shard planning, `.bin/.idx` creation, chat loss masks, packing algorithms, packed Parquet schema, and Bridge config integration.
   - Keep tests small; do not require large datasets or multi-node execution.

### Desired outcome

- Bridge documents raw input schemas for LLM pretraining and SFT/LoRA.
- Bridge documents prepared artifact contracts for `.bin/.idx` and packed Parquet.
- There is a clear mapping from prepared artifacts to Bridge recipe config fields.
- A toy pretraining dataset can be converted into `.bin/.idx` and consumed by a Bridge pretraining config.
- A toy SFT dataset can be converted into packed Parquet and consumed by a Bridge SFT/LoRA config.

### Affected area

area:data

### Urgency / use case

Important but not blocking
